src/state.rs

```diff
@@ -96,19 +96,27 @@ impl Contracts {
     }
 }
 
-/// The main network state.
+/// The main network state consists of three different states: `staged`, `head`
+/// and `origin`.
 ///
-/// It keeps two different states, the `origin` and the `head`. The `origin` is
-/// the starting state, and `head` is origin with all the received transactions
-/// applied.
+/// Any changes is applied to the `staged` state, and it's only committed to
+/// the `head` state when the [commit](`Self::commit`) method is called.
 ///
-/// It is possible to either [commit](`Self::commit`) to the `head` state,
-/// turning it into the new `origin`, or [reset](`Self::reset`) it back to
-/// `origin`.
+/// Calling the [push](`Self::push`) method will turn the `head` state into the
+/// new `origin`. At this point, any (`staged`) changes that are not committed
+/// will be discarded.
+///
+/// It's possible to manually discard the `staged` state by calling the
+/// [unstage](`Self::unstage`) method, as well as to reset all the three state
+/// to the `origin` by calling the [reset](`Self::reset`) method.
+///
+/// Due to the ephemeral nature of `staged` state, only `head` and `origin`
+/// are persisted on disk.
 #[derive(Clone, Default)]
 pub struct NetworkState {
-    origin: Contracts,
+    pub(crate) staged: Contracts,
```

Comment:
Having a `*_mut` that is private to the crate or having the field as private to the crate is literally equivalent, unless there is some logic involved and there isn't. Also, since it's private to the crate it doesn't impact the public API (e.g., if in the future we decided to rename the field, or adding / removing some state).

I didn't see any reason to have an additional method, also for the same reason you mentioned (it was used only in one single place).

Reply:
We had a chat and discussed how to remove the leaking of the current state (both by field access or method access) and ended up to open #305 as follow up issue to fix.